src/router.js

```diff
@@ -7,38 +7,53 @@ import { useContext, useMemo, useReducer, useLayoutEffect, useRef } from 'preact
  * @typedef {import('./internal.d.ts').VNode} VNode
  */
 
-let push, scope;
-const UPDATE = (state, url) => {
+/** @type {boolean} */
+let push;
+/** @type {string | RegExp | undefined} */
+let scope;
+
+/**
+ * @param {string} href
+ * @returns {boolean}
+ */
+function isInScope(href) {
+	return !scope || (typeof scope == 'string'
+		? href.startsWith(scope)
+		: scope.test(href)
+	);
+}
+
+/**
+ * @param {string} state
+ * @param {MouseEvent | PopStateEvent | { url: string, replace?: boolean }} action
```

Comment:
By no means does the following code pass TS's inspection, but it's at least easy to reason about. We can always sprinke in `instanceof` or `prop in action` checks later if really needed.